### PR TITLE
Add huggingface embeddings 

### PR DESCRIPTION
**Pull request by frillecode**
- [ ] I have battle-tested on Overtaci (RMAPPS1279)
- [ ] I have bumped the version as described in the [wiki](https://www.notion.so/Setting-up-Poetry-on-your-local-machine-76647d75e8fa4d9ba553cbc2a49946d9#dca1e0f2ce934a49acd1851aaf882d75)

### CODE REVIEW

**Comment by frillecode** [comment]
Den giver fejl fordi _sentence\_transformers_ library ikke er installeret. Har I en holdning til om jeg bør tilføje det eller om jeg skal omskrive koden, så den kun depender på _transformers_?

**Comment by HLasse** [comment]
Det er nok bedre at omskrive til ren torch/transformers. 
1. Færre dependencies (selvom sentence_transformers er lille)
2. Jeg er ikke sikker på at `SentenceTransformers` implementeringen håndterer modeller der _ikke_ er SentenceTransformers. Dvs. hvis vi bare vil have embeddings fra en e.g. `danish-bert` vil det her sandsynligvis breake.

Jeg læser koden igennem og kommer med forslag 👍

File: src/psycopmlutils/loaders/raw/load_text.py [hunk 1]
```diff
@@ -107,14 +109,70 @@ def _tfidf_featurize(
...
+        >>> huggingface_model_id = ("sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2")
+        >>> df_p = p / "synth_txt_data.csv"
+
+        >>> df = pd.read_csv(df_p)
+        >>> df = df.dropna()
+
+        >>> x = LoadText._huggingface_featurize(df, huggingface_model_id)
+    """
+
     # Load paraphrase-multilingual-MiniLM-L12-v2
```
Comment (HLasse) [inline src/psycopmlutils/loaders/raw/load_text.py:139]:
```suggestion
```

File: src/psycopmlutils/loaders/raw/load_text.py [hunk 2]
```diff
@@ -107,14 +109,70 @@ def _tfidf_featurize(
...
-    # split tokens to list of list if longer than allowed sequence length
-    # which is often 128 for sentence transformers
-    # encode tokens
-    # average by list of list
-    # return embeddings
-    raise NotImplementedError
+    model = SentenceTransformer(model_id)
```
Comment (HLasse) [inline src/psycopmlutils/loaders/raw/load_text.py:140]: Omskriv til ikke at bruge `SentenceTransformer`

File: src/psycopmlutils/loaders/raw/load_text.py [hunk 3]
```diff
@@ -107,14 +109,70 @@ def _tfidf_featurize(
...
+    seq_length = int(
+        model.max_seq_length / 1.5,
+    )  # allowing space for more word piece tokens than words in original sequence
+
+    # Generate embeddings
+    embeddings = []
+    for txt in text:
+        words = txt.split(" ")
+        # If text is not longer than allowed sequence length, extract and save embeddings
+        if len(words) <= seq_length:
```
Comment (HLasse) [inline src/psycopmlutils/loaders/raw/load_text.py:157]: I suggest rewriting this to be based on the sequence length of the tokenized data, following tokenization using the HF model. In this way we won't have to tokenize the document twice

File: src/psycopmlutils/loaders/raw/load_text.py [hunk 4]
```diff
@@ -107,14 +109,70 @@ def _tfidf_featurize(
...
+            embeddings.append(embd)
+        # If text is longer than allowed sequence length, split text into chunks before extracting embeddings and save average across chunks
+        else:
+            words_in_chunks = [
+                words[y - seq_length : y]
+                for y in range(seq_length, len(words) + seq_length, seq_length)
+            ]
+            chunks = [
+                " ".join(w) for w in words_in_chunks if len(w) == seq_length
+            ]  # drop small remainder of shorter size
```
Comment (HLasse) [inline src/psycopmlutils/loaders/raw/load_text.py:168]: Move this to a separate function (and write a test for it). Use more descriptive variable names than `y` and `w`. Also needs to be rewritten a bit to work with the tokenized input

**Review by HLasse — CHANGES_REQUESTED** [review]
Nice work Frida! 
A couple of things before merging:

1. Rewrite to pure torch/transformrs to make sure this works with different models (not just `SentenceTransformers`)
2. Extract your chunking logic to a loop and write a test for it
3. Update `load_synth_notes` to also take the HF featurizer as input
4. Write a test (in `tests/test_timeseriesflattener/test_add_values`) where you apply it on the synthetic data
5. Battle test it on Overtaci - check that it works on actual data

**Pull request closed without merging**